bin/pytorch_inference/CCommandParser.cc

```diff
@@ -0,0 +1,133 @@
+const std::string CCommandParser::VAR_ARG_PREFIX{"arg_"};
+
+CCommandParser::CCommandParser(std::istream& strmIn) : m_StrmIn(strmIn) {
+}
+
+bool CCommandParser::ioLoop(const TRequestHandlerFunc& requestHandler) const {
+
+    rapidjson::IStreamWrapper isw(m_StrmIn);
+
+    while (true) {
+        rapidjson::Document doc;
+        rapidjson::ParseResult parseResult =
+            doc.ParseStream<rapidjson::kParseStopWhenDoneFlag>(isw);
+
+        if (static_cast<bool>(parseResult) == false) {
+            if (m_StrmIn.eof()) {
+                break;
+            }
+
+            LOG_ERROR(<< "Error parsing command from JSON: "
+                      << rapidjson::GetParseError_En(parseResult.Code())
+                      << ". At offset: " << parseResult.Offset());
+
+            return false;
+        }
+
+        if (validateJson(doc) == false) {
+            continue;
+        }
+
+        // TODO if logger.trace_enabled then
+        debug(doc);
+        CCommandParser::SRequest request = jsonToRequest(doc);
+        if (requestHandler(request) == false) {
+            LOG_ERROR(<< "Request handler forced exit");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool CCommandParser::validateJson(const rapidjson::Document& doc) const {
+    if (doc.HasMember(REQUEST_ID) == false) {
+        LOG_ERROR(<< "Invalid command: missing field [" << REQUEST_ID << "]");
+        return false;
+    }
+
+    if (doc.HasMember(TOKENS) == false) {
+        LOG_ERROR(<< "Invalid command: missing field [" << TOKENS << "]");
+        return false;
+    }
+
+    const rapidjson::Value& tokens = doc[TOKENS];
+    if (tokens.IsArray() == false) {
+        LOG_ERROR(<< "Invalid command: expected an array [" << TOKENS << "]");
+        return false;
+    }
+
+    // check optional args
+    std::uint64_t varCount{1};
+    std::string varArgName = VAR_ARG_PREFIX + std::to_string(varCount);
+    while (doc.HasMember(varArgName)) {
+        const rapidjson::Value& value = doc[varArgName];
+        if (value.IsArray() == false) {
+            LOG_ERROR(<< "Invalid command: argument [" << varArgName << "] is not an array");
+            return false;
+        }
+
+        ++varCount;
+        varArgName = VAR_ARG_PREFIX + std::to_string(varCount);
+    }
+
+    return true;
+}
+
+CCommandParser::SRequest CCommandParser::jsonToRequest(const rapidjson::Document& doc) const {
+    TUint32Vec tokens;
+    const rapidjson::Value& arr = doc[TOKENS];
+    for (auto itr = arr.Begin(); itr != arr.End(); ++itr) {
+        tokens.push_back(itr->GetUint());
+    }
+
+    std::uint64_t varCount{1};
+    std::string varArgName = VAR_ARG_PREFIX + std::to_string(varCount);
+    TUint32VecVec args;
+
+    while (doc.HasMember(varArgName)) {
+        TUint32Vec arg;
+        const rapidjson::Value& v = doc[varArgName];
+        for (auto itr = v.Begin(); itr != v.End(); ++itr) {
+            arg.push_back(itr->GetUint());
```

Comment:
I wonder if we should be checking the type is int here since this is effectively unknown input.

Reply:
This is not user input it will be generated by Elasticsearch, in fact I was tempted to remove all validation as we control the input. If the JSON string parses it can be considered well formed, only a catastrophic failure such as a broken pipe or simultaneous writes would invalidate the input.

Reply:
I did think quite a lot about the validation considered copying the pattern used for AD job configuration but felt that was unnecessary complex for this relatively simple document.

Reply:
I think this warrants a comment, because it isn't clear in this context that this would never be under a user control. (It is also worth considering whether it could be under a users control if abused.)

Reply:
I'd missed the general comment in the class description, that's good enough.